### ISSUE TITLE
Ryzen 1700 test failures

### ISSUE THREAD

**Issue by bernhardu** [mozilla/rr#2687]
Tried to run the tests on a `AMD Ryzen 7 1700 Eight-Core Processor (family: 0x17, model: 0x1, stepping: 0x1)` with the `nospec_store_bypass_disable` workaround, running this stock debian kernel `Linux rechner 5.8.0-1-amd64 #1 SMP Debian 5.8.7-1 (2020-09-05) x86_64 GNU/Linux`.

I still had to disable this [fprintf](https://github.com/mozilla/rr/blob/fe2c788be55a4d147c69641bd07f9d686fce25f3/src/PerfCounters_x86.h#L287), otherwise all tests fail.
Therefore I don't know if the `nospec_store_bypass_disable` works as expected, at least `/proc/self/status` shows `Speculation_Store_Bypass:       vulnerable`

```
$ git log --oneline | head -n1
547a3166 Handle legacy getresuid/getresgid not present
$ cd ../obj
$ ctest -j$(nproc)
...
99% tests passed, 9 tests failed out of 2487
...
The following tests FAILED:
        436 - x86/ptrace_debug_regs (Failed)
        437 - x86/ptrace_debug_regs-no-syscallbuf (Failed)
        458 - ptrace_sysemu (Failed)
        818 - conditional_breakpoint_offload (Failed)
        1680 - x86/ptrace_debug_regs-32 (Failed)
        1681 - x86/ptrace_debug_regs-32-no-syscallbuf (Failed)
        1703 - ptrace_sysemu-32-no-syscallbuf (Failed)
        2061 - conditional_breakpoint_offload-32-no-syscallbuf (Failed)
        2394 - record_replay-32 (Failed)
Errors while running CTest
```
On three other runs there were a few more failures.
I would be happy if I could help by doing some more testing.

Also I tried to repeat this test inside a qemu VM with `-cpu host`, but there
I receive an error `unchecked MSR access error: WRMSR to 0xc0010200` in dmesg and most tests fail.

**Comment by khuey** [comment]
The `ptrace_debug_regs` failures are #2636 
Can you post the .err files from running `ptrace_sysemu`?

**Comment by bernhardu** [comment]
Hello, I reran the test with `ctest --verbose --tests-regex "^ptrace_sysemu$"` and received this  [replay.err.gz](https://github.com/mozilla/rr/files/5250063/replay.err.gz), the record.err was empty.

I looked through my /tmp and found also some which have [record.err.gz](https://github.com/mozilla/rr/files/5250066/record.err.gz). Here the replay.err was empty.

**Comment by khuey** [comment]
Ok rr is stuck waiting on the replay tracee. The replay tracee is supposedly running which is fascinating since there are no loops or anything in that test.

If you `rr replay -a <whatever directory you found that replay.err in>/latest-trace` does it also hang?

**Comment by bernhardu** [comment]
The first 4 attempts just wrote `EXIT-SUCCESS`, the fifth did then hang:
`$ bin/rr replay -a /tmp/rr-test-ptrace_sysemu-vjBo5FR7G/latest-trace`

```
$ pstree -p
...
           │                ├─konsole(2641)─┬─bash(2666)───script(387043)───bash(387044)───rr(519403)───rr:ptrace_sysem(519404)───rr:ptrace_sysem(519405)
...
$ gdb -q --pid 519403
Attaching to process 519403
...
Reading symbols from /usr/lib/debug/.build-id/63/7706dbbbd112d03fbad61ca30125b48e60aa92.debug...
0x00007fd89df5d339 in __waitid (idtype=P_PID, id=519405, infop=0x7ffec2fb2380, options=2) at ../sysdeps/unix/sysv/linux/waitid.c:29
29      ../sysdeps/unix/sysv/linux/waitid.c: Datei oder Verzeichnis nicht gefunden.
(gdb) set width 0
(gdb) bt
#0  0x00007fd89df5d339 in __waitid (idtype=P_PID, id=519405, infop=0x7ffec2fb2380, options=2) at ../sysdeps/unix/sysv/linux/waitid.c:29
#1  0x00005622898e8596 in rr::Task::wait (this=0x56228b420c80, interrupt_after_elapsed=-1) at /home/bernhard/data/entwicklung/2020/rr/2020-09-19-amd64/rr/src/Task.cc:1842
#2  0x00005622898e6dcd in rr::Task::resume_execution (this=0x56228b420c80, how=rr::RESUME_SYSEMU, wait_how=rr::RESUME_WAIT, tick_period=rr::RESUME_UNLIMITED_TICKS, sig=0) at /home/bernhard/data/entwicklung/2020/rr/2020-09-19-amd64/rr/src/Task.cc:1435
#3  0x000056228987c66c in rr::ReplaySession::cont_syscall_boundary (this=0x56228b4069e0, t=0x56228b420c80, constraints=...) at /home/bernhard/data/entwicklung/2020/rr/2020-09-19-amd64/rr/src/ReplaySession.cc:462
#4  0x000056228987cefc in rr::ReplaySession::enter_syscall (this=0x56228b4069e0, t=0x56228b420c80, constraints=...) at /home/bernhard/data/entwicklung/2020/rr/2020-09-19-amd64/rr/src/ReplaySession.cc:581
#5  0x0000562289880f1c in rr::ReplaySession::try_one_trace_step (this=0x56228b4069e0, t=0x56228b420c80, constraints=...) at /home/bernhard/data/entwicklung/2020/rr/2020-09-19-amd64/rr/src/ReplaySession.cc:1409
#6  0x00005622898822b4 in rr::ReplaySession::replay_step (this=0x56228b4069e0, constraints=...) at /home/bernhard/data/entwicklung/2020/rr/2020-09-19-amd64/rr/src/ReplaySession.cc:1687
#7  0x0000562289879672 in rr::ReplaySession::replay_step (this=0x56228b4069e0, command=rr::RUN_CONTINUE) at /home/bernhard/data/entwicklung/2020/rr/2020-09-19-amd64/rr/src/ReplaySession.h:284
#8  0x0000562289878100 in rr::serve_replay_no_debugger (trace_dir="/tmp/rr-test-ptrace_sysemu-vjBo5FR7G/latest-trace", flags=...) at /home/bernhard/data/entwicklung/2020/rr/2020-09-19-amd64/rr/src/ReplayCommand.cc:362
#9  0x0000562289878632 in rr::replay (trace_dir="/tmp/rr-test-ptrace_sysemu-vjBo5FR7G/latest-trace", flags=...) at /home/bernhard/data/entwicklung/2020/rr/2020-09-19-amd64/rr/src/ReplayCommand.cc:450
#10 0x000056228987936a in rr::ReplayCommand::run (this=0x562289b167a0 <rr::ReplayCommand::singleton>, args=std::vector of length 0, capacity 4) at /home/bernhard/data/entwicklung/2020/rr/2020-09-19-amd64/rr/src/ReplayCommand.cc:616
#11 0x000056228993a611 in main (argc=4, argv=0x7ffec2fb3698) at /home/bernhard/data/entwicklung/2020/rr/2020-09-19-amd64/rr/src/main.cc:268
```

**Comment by khuey** [comment]
When that happens does the /proc/<pid>/status for the `rr:ptrace_sysemu` process say `State: R (running)`?

**Comment by khuey** [comment]
Also if you could run with `RR_LOG=all:debug` and catch a hanging run, and then provide that log, that would be great.

**Comment by bernhardu** [comment]
For some reason when I came back to that system the hanging process finished ...

Another hanging attempt of `$ bin/rr replay -a /tmp/rr-test-ptrace_sysemu-vjBo5FR7G/latest-trace` showed:
```
...
           |                |-konsole(2641)-+-bash(2666)---script(387043)---bash(387044)---rr(528812)---rr:ptrace_sysem(528813)---r+
...

root@rechner:~# cat /proc/528812/status
Name:   rr
Umask:  0022
State:  S (sleeping)
...
root@rechner:~# cat /proc/528813/status
Name:   rr:ptrace_sysem
Umask:  0022
State:  t (tracing stop)
```

**Comment by khuey** [comment]
Ok, that makes more sense than a running state.

Can you get that RR_LOG=all:debug log?

**Comment by bernhardu** [comment]
A `script -c "RR_LOG=all:debug bin/rr replay -a /tmp/rr-test-ptrace_sysemu-vjBo5FR7G/latest-trace" -a ../rr_$(date +%Y-%m-%d_%H-%M-%S).log` catched follogwin output until I did a `kill 529257`:
[rr_2020-09-20_04-14-58.log](https://github.com/mozilla/rr/files/5251061/rr_2020-09-20_04-14-58.log)

**Comment by bernhardu** [comment]
On todays boot I selected the previous kernel version with nospec_store_bypass_disable and built current rr git.
Running the test suite on that gave just two failing tests, which did also succeed on running them specifically.
```
$ (cd ../rr; git log --oneline | head -n1)
a4f96c01 Allow SIGKILL to be returned from unexpected_exit_pid_ns

$ uname -a
Linux rechner 5.7.0-3-amd64 #1 SMP Debian 5.7.17-1 (2020-08-23) x86_64 GNU/Linux
bernhard@rechner:~/data/entwicklung/2020/rr/2020-09-21-amd64/obj$ ctest -j$(nproc)
...
99% tests passed, 2 tests failed out of 2487

Total Test time (real) = 548.88 sec

The following tests FAILED:
        819 - conditional_breakpoint_offload-no-syscallbuf (Failed)
        2061 - conditional_breakpoint_offload-32-no-syscallbuf (Failed)
Errors while running CTest
bernhard@rechner:~/data/entwicklung/2020/rr/2020-09-21-amd64/obj$ ctest --verbose --tests-regex "^conditional_breakpoint_offload-no-syscallbuf$"
...
1/1 Test #819: conditional_breakpoint_offload-no-syscallbuf ...   Passed   66.30 sec
...
bernhard@rechner:~/data/entwicklung/2020/rr/2020-09-21-amd64/obj$ ctest --verbose --tests-regex "^conditional_breakpoint_offload-32-no-syscallbuf$"
...
1/1 Test #2061: conditional_breakpoint_offload-32-no-syscallbuf ...   Passed   75.60 sec
...
```

**Comment by rocallahan** [comment]
Can you run ctest with `--verbose` on the entire suite and let me know what happens? Verbose logs for any failing tests will be useful.

**Comment by bernhardu** [comment]
@khuey : after yesterdays commit e12b61287 I lost the ability to replay. I receive following crash message:
```
[Di Sep 22 16:04:01 2020] rr[87168]: segfault at 98 ip 000055833903a25e sp 00007ffdb2d385a0 error 4 in rr[558338e9a000+29b000]
[Di Sep 22 16:04:01 2020] Code: 45 f8 8b 80 a4 00 00 00 5d c3 55 48 89 e5 48 89 7d f8 48 8b 45 f8 8b 80 a8 00 00 00 5d c3 55 48 89 e5 48 89 7d f8 48 8b 45 f8 <8b> 80 98 00 00 00 5d c3 55 48 89 e5 48 83 ec 20 48 89 7d e8 48 8d
```
When I make following change it starts to work again:
```
-          << "Expected syscall_bp_vm to be clear but it's " << syscall_bp_vm->leader_tid()
+          << "Expected syscall_bp_vm to be clear but it's " //<< syscall_bp_vm->leader_tid()
```

**Comment by khuey** [comment]
Thanks, I've been writing too much Rust lately. Pull the latest tip.

**Comment by bernhardu** [comment]
@khuey : I don't see a followup change related to e12b612. Got there something pushed to git?

**Comment by khuey** [comment]
Sorry I only pushed it to my personal repo. Try again.

**Comment by bernhardu** [comment]
```
$ script -c "(cd ../rr; git log --oneline | head -n1); uname -a; ctest --verbose -j1" -a "../rr-ctest_$(date +%Y-%m-%d_%H-%M-%S).log"
...
3d85311d Fix assertion.
Linux rechner 5.8.0-2-amd64 #1 SMP Debian 5.8.10-1 (2020-09-19) x86_64 GNU/Linux
...
100% tests passed, 0 tests failed out of 2487

Total Test time (real) = 3993.16 sec
```
I ran the tests today again and all tests did succeed (with nospec_store_bypass_disable).

**Comment by rocallahan** [comment]
Looks like we're done here?

**Comment by bernhardu** [comment]
@rocallahan : I guess yes.

But just today test `clone_interruption-32-no-syscallbuf` failed three of three times.
[rr-ctest_2020-09-24_16-54-08.log.gz](https://github.com/mozilla/rr/files/5278571/rr-ctest_2020-09-24_16-54-08.log.gz)

**Comment by rocallahan** [comment]
For a failing run, can you follow the instructions like this and tell us what happens?
```
_RR_TRACE_DIR=/tmp/rr-test-clone_interruption_32-vZtIbXl7g rr replay
```

**Comment by rocallahan** [comment]
er hmmm
```
2053: args: [b'/home/bernhard/data/entwicklung/2020/rr/2020-09-24-amd64/obj/bin/rr', b'--suppress-environment-warnings', 
b'--check-cached-mmaps', b'--fatal-errors', b'--resource-path=/home/bernhard/data/entwicklung/2020/rr/2020-09-24-amd64/ob
j', b'replay', b'-o-n', b'-x', b'/home/bernhard/data/entwicklung/2020/rr/2020-09-24-amd64/rr/src/test/test_setup.gdb', b'
-g', b'999999999']
```
That event number there looks bogus

**Comment by rocallahan** [comment]
We didn't find `SIGNAL` in the trace :-(. Can you get an `rr dump` of a failing trace and make it accessible please?

**Comment by bernhardu** [comment]
[rr-replay.txt](https://github.com/mozilla/rr/files/5280522/rr-replay.txt)
[rr-dump_2020-09-25_05-55-16.log.gz](https://github.com/mozilla/rr/files/5280517/rr-dump_2020-09-25_05-55-16.log.gz)

**Comment by rocallahan** [comment]
OK, it's a timeout. The test just hangs during recording:
```
{
  real_time:6472.669360 global_time:2198, event:`SYSCALL: read' (state:ENTERING_SYSCALL) tid:116600, ticks:58674
eax:0xffffffda ecx:0xff938b07 edx:0x1 ebx:0x4 esp:0xff938a4c ebp:0xff938b18 esi:0xf7f71000 edi:0x0 eip:0xf7fc6572 eflags:0x246 xcs:0x23 xss:0x2b xds:0x2b xes:0x2b xfs:0x0 xgs:0x63 orig_eax:0x3
}
{
  real_time:6594.605371 global_time:2199, event:`EXIT' tid:116601, ticks:490
}
{
  real_time:6594.605824 global_time:2200, event:`EXIT' tid:116600, ticks:58674
}
```

**Comment by rocallahan** [comment]
Looks like we sent `SIGCHLD` to the child thread and it just ... disappeared.

**Comment by rocallahan** [comment]
@bernhardu if you record that test many times and find a stuck one, dumping /proc/<pid>/status would be useful.

**Comment by rocallahan** [comment]
A stuck recording, that is.

**Comment by bernhardu** [comment]
Sorry for the delay. Unfortunately after a reboot the clone_interruption-32-no-syscallbuf succeeded in 100 runs.
So all tests pass (conditional_breakpoint_offload* on a second, non-parallel attempt).
Except the new x86/x87env test.
One thing I found about that test is that stepping in plain gdb with nexti through the test fails the same way.

**Comment by rocallahan** [comment]
Can you describe the x87env test failure in greater detail? What gets printed?

**Comment by bernhardu** [comment]
[rr-ctest_x86-x87env_2020-09-26_09-34-19.log](https://github.com/mozilla/rr/files/5286336/rr-ctest_x86-x87env_2020-09-26_09-34-19.log)

**Comment by bernhardu** [comment]
But unlike yesterday the test binary executed alone fails today that way:
```
bernhard@rechner:~/data/entwicklung/2020/rr/2020-09-26-amd64/obj$ /tmp/rr-test-x87env-FvAhLNv2r/x87env-FvAhLNv2r
FIP=0x99e9f35c
XSAVE FIP=0x0
FAILED: errno=0 (Success)
x87env-FvAhLNv2r: /home/bernhard/data/entwicklung/2020/rr/2020-09-26-amd64/rr/src/test/x86/x87env.c:29: main: Assertion `"FAILED: !" && check_cond(xsave_buf[1] == (uintptr_t)&_x87_instruction)' failed.
Abgebrochen (Speicherabzug geschrieben)
```

**Comment by rocallahan** [comment]
Excellent! This basically tells me what I needed to know for https://github.com/mozilla/rr/issues/2698.

This happens every time I suppose?

**Comment by bernhardu** [comment]
Yes, in 20 attempts all fail. Running the test with "FIP=0x0", running the plain binary with "XSAVE FIP=0x0".

**Comment by rocallahan** [comment]
You mean that when you `rr record x87env`, it fails with `FIP=0x0`?

**Comment by rocallahan** [comment]
But if you revert to revision b7884be64599eddc4c157629148ca655d476e3c0, does `rr record x87env` run OK?

**Comment by rocallahan** [comment]
can you give the exact output of the test run with both revisions, both under rr and not under rr?

**Comment by bernhardu** [comment]
Yes:
```
bernhard@rechner:~/data/entwicklung/2020/rr/2020-09-26-amd64/obj$ (cd ../rr; git log -1 --oneline)
551b0029 (HEAD -> master, origin/master, origin/HEAD) Test XSAVE too
bernhard@rechner:~/data/entwicklung/2020/rr/2020-09-26-amd64/obj$ bin/rr ./bin/x87env
rr: Saving execution to trace directory `/home/bernhard/.local/share/rr/x87env-0'.
FIP=0x0
FAILED: errno=0 (Success)
x87env: /home/bernhard/data/entwicklung/2020/rr/2020-09-26-amd64/rr/src/test/x86/x87env.c:27: main: Assertion `"FAILED: !" && check_cond(fstenv_buf[3] == (uint32_t)(uintptr_t)&_x87_instruction)' failed.
Aborted
bernhard@rechner:~/data/entwicklung/2020/rr/2020-09-26-amd64/obj$ ./bin/x87env
FIP=0x205df35c
XSAVE FIP=0x0
FAILED: errno=0 (Success)
x87env: /home/bernhard/data/entwicklung/2020/rr/2020-09-26-amd64/rr/src/test/x86/x87env.c:29: main: Assertion `"FAILED: !" && check_cond(xsave_buf[1] == (uintptr_t)&_x87_instruction)' failed.
Aborted (core dumped)
```

But yesterday, before the XSAVE changes:
```
bernhard@rechner:~/data/entwicklung/2020/rr/2020-09-25-amd64/obj$ (cd ../rr; git log -1 --oneline)
f5d783df (HEAD -> master, origin/master, origin/HEAD) Fix flock64 definition for Arch Linux
bernhard@rechner:~/data/entwicklung/2020/rr/2020-09-25-amd64/obj$ bin/rr ./bin/x87env
rr: Saving execution to trace directory `/home/bernhard/.local/share/rr/x87env-1'.
FIP=0x0
FAILED: errno=0 (Success)
x87env: /home/bernhard/data/entwicklung/2020/rr/2020-09-25-amd64/rr/src/test/x86/x87env.c:23: main: Assertion `"FAILED: !" && check_cond(buf[3] == (uint32_t)(long)&_x87_instruction)' failed.
Abgebrochen
bernhard@rechner:~/data/entwicklung/2020/rr/2020-09-25-amd64/obj$ ./bin/x87env
FIP=0x6e1c7323
EXIT-SUCCESS
```

**Comment by rocallahan** [comment]
If you start the test with `gdb --args x87env`, `break main`, `continue`, `continue`, does the test pass or fail?

**Comment by bernhardu** [comment]
I got this:
```
bernhard@rechner:~/data/entwicklung/2020/rr/2020-09-26-amd64/obj$ (cd ../rr; git log -1 --oneline)
551b0029 (HEAD -> master, origin/master, origin/HEAD) Test XSAVE too
bernhard@rechner:~/data/entwicklung/2020/rr/2020-09-26-amd64/obj$ gdb -q --args ./bin/x87env
Reading symbols from ./bin/x87env...
(gdb) break main
Breakpoint 1 at 0x1333: file /home/bernhard/data/entwicklung/2020/rr/2020-09-26-amd64/rr/src/test/x86/x87env.c, line 12.
(gdb) run
Starting program: /home/bernhard/data/entwicklung/2020/rr/2020-09-26-amd64/obj/bin/x87env 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".

Breakpoint 1, main () at /home/bernhard/data/entwicklung/2020/rr/2020-09-26-amd64/rr/src/test/x86/x87env.c:12
12        uint64_t* xsave_buf = (uint64_t*)mmap(NULL, 65536, PROT_READ | PROT_WRITE, MAP_ANONYMOUS | MAP_PRIVATE, -1, 0);
(gdb) continue
Continuing.
FIP=0x5555535c
XSAVE FIP=0x0
FAILED: errno=0 (Success)
x87env: /home/bernhard/data/entwicklung/2020/rr/2020-09-26-amd64/rr/src/test/x86/x87env.c:29: main: Assertion `"FAILED: !" && check_cond(xsave_buf[1] == (uintptr_t)&_x87_instruction)' failed.

Program received signal SIGABRT, Aborted.
__GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
50      ../sysdeps/unix/sysv/linux/raise.c: No such file or directory.
```

**Comment by bernhardu** [comment]
I tried to set a single breakpoint to one debugged process and found these three with the fstenv failure, the others with the xsave:
```
script -a "../gdb_$(date +%Y-%m-%d_%H-%M-%S).log"

(cd ../rr; git log -1 --oneline)
gdb -q --args ./bin/x87env

display/i $pc
b main
run
disassemble main, main+134
b *main+38      FIP=0x5555535c  "FAILED: !" && check_cond(xsave_buf[1] == (uintptr_t)&_x87_instruction)' failed.
b *main+43      FIP=0x5555535c  "FAILED: !" && check_cond(xsave_buf[1] == (uintptr_t)&_x87_instruction)' failed.
b *main+48      FIP=0x5555535c  "FAILED: !" && check_cond(xsave_buf[1] == (uintptr_t)&_x87_instruction)' failed.
b *main+52      FIP=0x555550c2  x87env: .../x87env.c:27: main: Assertion `"FAILED: !" && check_cond(fstenv_buf[3] == (uint32_t)(uintptr_t)&_x87_instruction)' failed.
b *main+54      FIP=0x0         x87env: .../x87env.c:27: main: Assertion `"FAILED: !" && check_cond(fstenv_buf[3] == (uint32_t)(uintptr_t)&_x87_instruction)' failed.
b *main+59      FIP=0x5555535c  "FAILED: !" && check_cond(xsave_buf[1] == (uintptr_t)&_x87_instruction)' failed.
b *main+63      FIP=0x5555535c  "FAILED: !" && check_cond(xsave_buf[1] == (uintptr_t)&_x87_instruction)' failed.
b *main+73      FIP=0x0         x87env: .../x87env.c:27: main: Assertion `"FAILED: !" && check_cond(fstenv_buf[3] == (uint32_t)(uintptr_t)&_x87_instruction)' failed.
b *main+80      FIP=0x5555535c  "FAILED: !" && check_cond(xsave_buf[1] == (uintptr_t)&_x87_instruction)' failed.
b *main+84      FIP=0x5555535c  "FAILED: !" && check_cond(xsave_buf[1] == (uintptr_t)&_x87_instruction)' failed.
b *main+86      FIP=0x5555535c  "FAILED: !" && check_cond(xsave_buf[1] == (uintptr_t)&_x87_instruction)' failed.
b *main+92      FIP=0x5555535c  "FAILED: !" && check_cond(xsave_buf[1] == (uintptr_t)&_x87_instruction)' failed.
b *main+95      FIP=0x5555535c  "FAILED: !" && check_cond(xsave_buf[1] == (uintptr_t)&_x87_instruction)' failed.
b *main+98      FIP=0x5555535c  "FAILED: !" && check_cond(xsave_buf[1] == (uintptr_t)&_x87_instruction)' failed.
b *main+100     FIP=0x5555535c  "FAILED: !" && check_cond(xsave_buf[1] == (uintptr_t)&_x87_instruction)' failed.
b *main+102     FIP=0x5555535c  "FAILED: !" && check_cond(xsave_buf[1] == (uintptr_t)&_x87_instruction)' failed.
b *main+104     FIP=0x5555535c  "FAILED: !" && check_cond(xsave_buf[1] == (uintptr_t)&_x87_instruction)' failed.
b *main+107     FIP=0x5555535c  "FAILED: !" && check_cond(xsave_buf[1] == (uintptr_t)&_x87_instruction)' failed.
b *main+110     FIP=0x5555535c  "FAILED: !" && check_cond(xsave_buf[1] == (uintptr_t)&_x87_instruction)' failed.
b *main+115     FIP=0x5555535c  "FAILED: !" && check_cond(xsave_buf[1] == (uintptr_t)&_x87_instruction)' failed.
b *main+117     FIP=0x5555535c  "FAILED: !" && check_cond(xsave_buf[1] == (uintptr_t)&_x87_instruction)' failed.
b *main+119     FIP=0x5555535c  "FAILED: !" && check_cond(xsave_buf[1] == (uintptr_t)&_x87_instruction)' failed.
b *main+121     FIP=0x5555535c  "FAILED: !" && check_cond(xsave_buf[1] == (uintptr_t)&_x87_instruction)' failed.
b *main+124     FIP=0x5555535c  "FAILED: !" && check_cond(xsave_buf[1] == (uintptr_t)&_x87_instruction)' failed.
b *main+127     FIP=0x5555535c  "FAILED: !" && check_cond(xsave_buf[1] == (uintptr_t)&_x87_instruction)' failed.
b *main+129     FIP=0x5555535c  "FAILED: !" && check_cond(xsave_buf[1] == (uintptr_t)&_x87_instruction)' failed.
cont
cont
kill
y
q
exit

(gdb) disassemble main, main+134
Dump of assembler code from 0x555555555328 to 0x5555555553ae:
   0x0000555555555328 <main+0>: push   %rbp
   0x0000555555555329 <main+1>: mov    %rsp,%rbp
   0x000055555555532c <main+4>: sub    $0x420,%rsp
=> 0x0000555555555333 <main+11>:        mov    $0x0,%r9d
   0x0000555555555339 <main+17>:        mov    $0xffffffff,%r8d
   0x000055555555533f <main+23>:        mov    $0x22,%ecx
   0x0000555555555344 <main+28>:        mov    $0x3,%edx
   0x0000555555555349 <main+33>:        mov    $0x10000,%esi
   0x000055555555534e <main+38>:        mov    $0x0,%edi
   0x0000555555555353 <main+43>:        callq  0x555555555060 <mmap@plt>
   0x0000555555555358 <main+48>:        mov    %rax,-0x8(%rbp)
   0x000055555555535c <main+52>:        fld1   
   0x000055555555535e <main+54>:        callq  0x555555555080 <sched_yield@plt>
   0x0000555555555363 <main+59>:        mov    -0x8(%rbp),%rcx
   0x0000555555555367 <main+63>:        mov    $0xffffffff,%eax
   0x000055555555536c <main+68>:        mov    $0xffffffff,%edx
   0x0000555555555371 <main+73>:        fstenv -0x420(%rbp)
   0x0000555555555378 <main+80>:        xsave64 (%rcx)
   0x000055555555537c <main+84>:        fldz   
   0x000055555555537e <main+86>:        fldenv -0x420(%rbp)
   0x0000555555555384 <main+92>:        fstpt  -0x20(%rbp)
   0x0000555555555387 <main+95>:        fldt   -0x20(%rbp)
   0x000055555555538a <main+98>:        fld1   
   0x000055555555538c <main+100>:       fucomip %st(1),%st
   0x000055555555538e <main+102>:       fstp   %st(0)
   0x0000555555555390 <main+104>:       setnp  %al
   0x0000555555555393 <main+107>:       fldt   -0x20(%rbp)
   0x0000555555555396 <main+110>:       mov    $0x0,%edx
   0x000055555555539b <main+115>:       fld1   
   0x000055555555539d <main+117>:       fucomip %st(1),%st
   0x000055555555539f <main+119>:       fstp   %st(0)
   0x00005555555553a1 <main+121>:       cmovne %edx,%eax
   0x00005555555553a4 <main+124>:       movzbl %al,%eax
   0x00005555555553a7 <main+127>:       mov    %eax,%edi
   0x00005555555553a9 <main+129>:       callq  0x555555555298 <check_cond>
```
One thing that I want to mention is, if single stepping through the instructions the `fstenv` is executed in two steps:
```
(gdb) 
0x000055555555536c      17        __asm__ volatile("fstenv %1\n"
1: x/i $pc
=> 0x55555555536c <main+68>:    mov    $0xffffffff,%edx
(gdb) 
0x0000555555555371      17        __asm__ volatile("fstenv %1\n"
1: x/i $pc
=> 0x555555555371 <main+73>:    fstenv -0x420(%rbp)
(gdb) 
0x0000555555555372      17        __asm__ volatile("fstenv %1\n"
1: x/i $pc
=> 0x555555555372 <main+74>:    fnstenv -0x420(%rbp)
(gdb) 
0x0000555555555378      17        __asm__ volatile("fstenv %1\n"
1: x/i $pc
=> 0x555555555378 <main+80>:    xsave64 (%rcx)
```
[example for FIP=0x555550c2](https://github.com/mozilla/rr/files/5286953/gdb_2020-09-26_17-42-25.log)

**Comment by rocallahan** [comment]
What is at address 0x00005555555550c2? (i.e. disassembled)

**Comment by rocallahan** [comment]
(this information is very interesting, thanks!)

**Comment by rocallahan** [comment]
> What is at address 0x00005555555550c2? (i.e. disassembled)

Don't worry about it, it's strange, but the same thing seems to happen on Intel so I don't think it's related.

**Comment by rocallahan** [comment]
I will fix `x87env` for Zen in #2693 2698
